### ISSUE TITLE
load.image failed to load image from URL on windows

### ISSUE THREAD

**Issue by ShotaOchi** [dahtah/imager#64]
I tried to load a image from URL as shown below.
###########################################-
library(imager)
 a <- load.image("https://cran.r-project.org/Rlogo.jpg")
###########################################-

The code caused the following error for the first time.
------------------------------------------------------------------------------------------------------------------#
Error in load_image(file) : 
  cimg::stderr(): Reference to 'stderr' stream not allowed in R mode ('cimg_use_r' is defined).
------------------------------------------------------------------------------------------------------------------#

The code caused the following error from next time.
------------------------------------------------------------------------------------------------------------------#
Error in load_image(file) : 
  [instance(0,0,0,0,0000000000000000,non-shared)] CImg<double>::load(): Failed to recognize format of file 'C:\Users\user_name\AppData\Local\Temp\RtmpszXASx\file24346a0a7022'.
------------------------------------------------------------------------------------------------------------------#

The errors occurred on windows 10 and windows 7.
The errors didn't occur on Ubuntu.

The error from second time said that load_image failed to recognize the format of the temporary file.
So, a modification of wrap.url as shown below should fix it.
The modification appears in bold type.
###########################################-
wrap.url <- function(file,fun)
{
    is.url <- grepl("^(http|ftp)s?://", file)
    if (is.url)
    {
        url <- file
       **ext <- gsub(".*\\.", ".",  url)** 
        file <- tempfile(**fileext = ext**)
        downloader::download(url,file)
###########################################-

I don't understand how to fix the error for the first time.
It seems that cimg tried to open the standard error stream.
But I have no idea how to fix the error.

**Comment by dahtah** [comment]
Thanks for reporting this - can you please give me a few more details? The first message is due to some bad error handling in the interface with CImg, which I should fix. What happens is that CImg tries to write to stderr, which is disallowed by CRAN, and instead of getting a helpful error message you get that obscure error message. So something happened when trying to load the image but we don't know what. My guess is that it tried to open the image via imagemagick, which may not be on your system. 
I've integrated your fix to wrap_url, it's a very good idea. Can you test again with the current master? Do you need a binary package for Windows?

**Comment by ShotaOchi** [comment]
Thanks.
load.image function returned the following message from the first time. 
------------------------------------------------------------------------------------------------------------------#
JPEG decompression: Corrupt JPEG data: 1 extraneous bytes before marker 0xd9Error in readfun(f, ...) : 
  JPEG decompression error: JPEG datastream contains no image
------------------------------------------------------------------------------------------------------------------#

The corruption was because of download function.
I confirmed download function caused the corruption by the code shown below.
#############################################################################-
library(imager)
library(downloader)
url <- "https://cran.r-project.org/Rlogo.jpg"
temp1 <- tempfile()
temp2 <- tempfile()
download(url, temp1)
download(url, temp2, **mode = "wb"**)

load.image(temp1)
load.image(temp2)
#############################################################################-

load.image(temp1) returned the following message.
------------------------------------------------------------------------------------------------------------------#
Error in load_image(file) : 
  cimg::stderr(): Reference to 'stderr' stream not allowed in R mode ('cimg_use_r' is defined).
------------------------------------------------------------------------------------------------------------------#

load.image(temp2) returned the following message.
------------------------------------------------------------------------------------------------------------------#
Image. Width: 100 pix Height: 76 pix Depth: 1 Colour channels: 3 
------------------------------------------------------------------------------------------------------------------#

load.image(temp2) worked correctly! 
I confirmed it by plotting the result of load.image(temp2).

To sum up, transferring the file in ASCII mode caused the error.
We should trasfer the file in Binary mode.
That's why the modification of wrap.url function shown below is enough.
The modification appears in bold type.
#############################################################################-
wrap.url <- function(file,fun)
{
is.url <- grepl("^(http|ftp)s?://", file)
if (is.url)
{
        url <- file
        ext <- stringr::str_extract_all(url,"\\.([A-Za-z0-9]+$)")[[1]]
        if (length(ext) > 0) file <- tempfile(fileext=ext) else file <- tempfile()
downloader::download(url,file, **mode = "wb"**)
#############################################################################-

**Comment by dahtah** [comment]
Thanks, fix implemented! It's in the master branch, Win binary here: 
https://win-builder.r-project.org/9lFH1xOik9Mm
